plugins/net_plugin/net_plugin.cpp

```diff
@@ -1165,8 +1164,8 @@ namespace eosio {
    }
 
    // called from connection strand
-   void connection::check_heartbeat( tstamp current_time ) {
-      if( latest_msg_time > 0 ) {
+   void connection::check_heartbeat( std::chrono::system_clock::time_point current_time ) {
+      if( latest_msg_time > std::chrono::system_clock::time_point::min() ) {
          if( current_time > latest_msg_time + hb_timeout ) {
```

Comment:
The issue was here. `current_time` was in nanoseconds. `lastest_msg_time` was also in nanoseconds, but `hb_timeout` was in `milliseconds`. So when the timer ran the `latest_msg_time` had to be within 10000 nanoseconds (10 microseconds) or it would close the connection.

Reply:
So the issue was that `hb_timeout` was in milliseconds. It seems incorrect to me, as it is defined as `std::chrono::system_clock::duration::rep`, so I feel we should consistently use `system_clock` duration in tstamp (on my linux system, `system_clock`'s `duration` is defined as `chrono::nanoseconds`.

I assume that by using `std::chrono` types, we will benefit from an automatic conversion in `if( current_time > latest_msg_time + hb_timeout )`  which will prevent the error. But `tstamp` was already a std::chrono duration, so I wonder if most of the change are really needed.

My feeling is that the change at line 610 (`hb_timeout = msec;`) would have been enough for the fix, as the problem was that we used `count()` which just returned a tick count.

Reply:
Could be. Either way, this is much cleaner/safer code.

Reply:
I don't see how this makes the code cleaner or safer. I think if you changed line 610 to `hb_timeout = msec.count()` the issue would come back exactly as before.

Reply:
I'm confused at what you are suggesting should be done. `time_point`, `milliseconds`, `nanoseconds` types are well defined and arithmetic operations are handled correctly. `tstamp` (`std::chrono::system_clock::duration::rep`) on the other hand is defined as `rep` - `signed arithmetic type representing the number of ticks in the clock's duration` which is system dependent.

Reply:
Is nanoseconds too sensitive in P2P?

Reply:
Yes, nanoseconds is kind of ridiculous to use. However, that is what we were already using for `time_message` and `time` of `handshake_message`. Internally now with this change we just use whatever `std::chrono::system_clock::now()` returns.

Reply:
Completely removed `tstamp` type and backported the better `time_message` handling from `main`.

Reply:
Thanks Kevin for the change. I'm not quite sure is the initialization with `std::chrono::system_clock::time_point::min()` has a specific benefit, I kinda feel  that `0` is more appropriate than a negative value. And I also thing that defining `tstamp` as `std::chrono::nanoseconds` and using it for our `connection` time tracking members would be better. But maybe we can talk about this further for 5.0.